cloud-controller-manager/do/loadbalancers.go

```diff
@@ -866,11 +872,41 @@ func buildForwardingRule(service *v1.Service, port *v1.ServicePort, protocol, ce
 	return &forwardingRule, nil
 }
 
-func buildFirewall(service *v1.Service) *godo.LBFirewall {
+func buildFirewall(service *v1.Service) (*godo.LBFirewall, error) {
+	var allow []string
+	sourceRanges, err := getSourceRangeRules(service)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get source range rules: %s", err)
+	}
+	if len(sourceRanges) > 0 {
+		// source ranges should take precedence over annotation based allow rules
+		allow = sourceRanges
+	} else {
+		allow = getStrings(service, annDOAllowRules)
+	}
 	return &godo.LBFirewall{
 		Deny:  getStrings(service, annDODenyRules),
-		Allow: getStrings(service, annDOAllowRules),
+		Allow: allow,
+	}, nil
+}
+
+// getSourceRangeRules returns loadBalancerSourceRanges in the format
+// `cidr:{source}`
+func getSourceRangeRules(service *v1.Service) ([]string, error) {
+	if len(service.Spec.LoadBalancerSourceRanges) == 0 {
+		return nil, nil
+	}
+	specs := service.Spec.LoadBalancerSourceRanges
+	ipnets, err := utilnet.ParseIPNets(specs...)
+	if err != nil {
+		return nil, fmt.Errorf("service.Spec.LoadBalancerSourceRanges: %v is not valid. Expecting a list of IP ranges. For example, 10.0.0.0/24. Error msg: %v", specs, err)
```

Comment:
This [can be stuttering](https://go.dev/play/p/p8c7lDjyRT9) since `net.ParseCIDRs` (which `ParseIPNets` eventually calls) may already be talking about addresses being invalid.

Can we address that and also be more concise, e.g.:

```golang
fmt.Errorf("failed to parse service.Spec.LoadBalancerSourceRanges %q: %s (expected format for an IP address is 10.0.0.0/24)", specs, err)
```